### ISSUE TITLE
Suffixed string and char literals are accepted if they don't go through parser

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#60494]
Things like `"abc"suffix` are lexically valid tokens, but they are reported as errors when parsed with the Rust Language Parser.

However, macro inputs are not parsed by the Rust Language Parser, each macro has its own language that it parses by itself.

It means that suffixed string and char literals can be passed to procedural macros (a procedural macro can process them as they want and make them legal and meaningful in its language), or declarative macros (if they are ignored).

Example:
```rust
macro_rules! blackhole { ($tt:tt) => () }

fn main() {
    blackhole!("string"suffix); // OK
}
```

Possible solutions:
- Alternative 1: Prohibit this during lexing, a string/character literal tokens won't be able to be suffixed, proc macros won't be able to use them in their DSLs.
- Alternative 2: Do nothing, string/character literal tokens can be suffixed (but not in the Rust Language proper), proc macros will be able to use them in their DSLs.

**Comment by petrochenkov** [comment]
This is listed as an unresolved question in the RFC introducing literal suffixes into lexer (https://github.com/rust-lang/rfcs/pull/463).
@nrc argued for the currently implemented behavior in https://github.com/rust-lang/rfcs/pull/463#discussion_r20414904.

**Comment by petrochenkov** [comment]
@rust-lang/lang needs to make some decision and document it in the [reference](https://github.com/rust-lang-nursery/reference/blob/master/src/tokens.md).

**Comment by Centril** [comment]
@petrochenkov What are the advantages to Alt 1. as you see it? Also, do we have some sort of crater data?

**Comment by petrochenkov** [comment]
>What are the advantages to Alt 1. as you see it?

This is just a somewhat exceptional case of a token which is lexically valid, but always rejected by parser.
I think `~` is the only other example of such token right now (and also `<-`, but that's only because const generics like `Type<-1>` are not properly parsed yet).

After some thought, yeah, there's probably not much advantage.

**Comment by Centril** [comment]
Yeah my inclination is that this may be useful for macro authors somehow so I think we can leave this accepted.

**Comment by eddyb** [comment]
I regret not coming up with the "joint" system for composed tokens much earlier, and doing *at least* literal suffixes with it. But overall, I'm in favor of as much flexibility as possible for proc macro authors.

The only *useful* constraint I see, in terms of the token model, is having balanced `()`/`[]`/`{}`, because they are used in the invocation syntax, and therefore allow unambiguous parsing around the invocation.
String literals and comments must have a fixed grammar *specifically* because they can contain those characters, and knowing whether it's part of the source or part of a literal/comment is crucial.

(Although I don't know if other @rust-lang/lang members share this view)

**Comment by nikomatsakis** [comment]
I agree with @eddyb as well =)

**Comment by petrochenkov** [comment]
Looks like there's some consensus to keep the existing behavior.
I made a PR to the reference documenting it - https://github.com/rust-lang-nursery/reference/pull/612.
